module/VuFind/src/VuFind/I18n/Initializer.php

```diff
@@ -0,0 +1,289 @@
+use Zend\Http\PhpEnvironment\Request;
+use Zend\I18n\Translator\Translator as I18nTranslator;
+use Zend\Mvc\I18n\Translator as MvcTranslator;
+use Zend\View\Model\ViewModel;
+
+/**
+ * Handles I18n initialization.
+ *
+ * @category VuFind
+ * @package  Translator
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Sebastian Kehr <kehr@ub.uni-leipzig.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class Initializer
+{
+    /**
+     * Translator.
+     *
+     * @var I18nTranslator
+     */
+    protected $translator;
+
+    /**
+     * Language base directories.
+     *
+     * @var string[]
+     */
+    protected $baseDirs = [];
+
+    /**
+     * File loader.
+     *
+     * @var ExtendedIni
+     */
+    protected $loader;
+
+    /**
+     * Initializer constructor.
+     *
+     * @param Request          $request       Request
+     * @param Config           $config        Configuration
+     * @param CookieManager    $cookies       Cookie manager
+     * @param StorageInterface $cache         Language cache
+     * @param MvcTranslator    $mvcTranslator Translator
+     * @param ViewModel        $viewModel     ViewModel
+     */
+    public function __construct(
+        Request $request,
+        Config $config,
+        CookieManager $cookies,
+        StorageInterface $cache,
+        MvcTranslator $mvcTranslator,
+        ViewModel $viewModel
+    ) {
+        $language = $this->detectLanguage($config, $request, $cookies);
+
+        $this->translator = $mvcTranslator->getTranslator();
+        $this->loader = $this->translator->getPluginManager()
+            ->get(ExtendedIni::class);
+
+        $this->translator->setLocale($language);
+        $this->translator->setCache($cache);
+
+        $this->loader->setFallbacks($this->parseFallbacks($config));
+
+        $rtlLangs = isset($config->LanguageSettings->rtl_langs)
+            ? array_map(
+                'trim', explode(',', $config->LanguageSettings->rtl_langs)
+            ) : [];
+
+        $viewModel->setVariable('userLang', $language);
+        $viewModel->setVariable('allLangs', $config->Languages);
+        $viewModel->setVariable('rtl', in_array($language, $rtlLangs));
+
+        $this->addBaseDir(APPLICATION_PATH . '/languages');
+        $this->addBaseDir(LOCAL_OVERRIDE_DIR . '/languages');
+    }
+
+    /**
+     * Initializes the translator and loader.
+     *
+     * @return void
+     */
+    public function init()
+    {
+        $cache = $this->translator->getCache();
+        $cacheKey = md5(implode('', $this->baseDirs));
+
+        if (!$cache->hasItem($cacheKey)) {
+            $cache->setItem($cacheKey, $this->detectTextDomains());
+        }
+
+        foreach ($cache->getItem($cacheKey) as $textDomain) {
+            $this->translator->addRemoteTranslations(
+                ExtendedIni::class, $textDomain
+            );
+        }
```

Comment:
Text domains will now be detected in all added base directories - but only once due to being cached.

Moreover translations are now added as remote translations as this better abstracts away possible locations of language files. (Also see comment on [`VuFind\I18n\Translator\Loader\ExtendedIni`](https://github.com/vufind-org/vufind/blob/c5ec6f98ac86daef4200b50d3d65bbbd495371d4/module/VuFind/src/VuFind/I18n/Translator/Loader/ExtendedIni.php).)

Reply:
Makes sense -- I was not aware of this distinction before, but it is useful.

Reply:
Speaking of caching, what is the cache lifetime? I've noticed that the cached translations seem to live basically forever; it's been in the back of my mind to investigate why and whether there's a way to better break the cache when things change... but I haven't found the time to deal with it since it's not a high-priority problem. Just wondering if your close investigation of all things i18n has revealed a possible better approach.

Reply:
Yes, filesystem caches as currently available/created by [`VuFind\Cache\Manager::getCache()`](https://github.com/vufind-org/vufind/blob/d4d62730859eaebbcda28ae51e1c57e4641f6e39/module/VuFind/src/VuFind/Cache/Manager.php#L131) will by default have an infinite TTL for stored items (corresponding to a value of 0).

Luckily this value can be modified for any cache instance: `$cache->getOptions()->setTtl($ttl)`.

Yet, as I just found out, setting a TTL for cached translations is currently not as easy as the cache instance actually used by the translator service depends on the enabled theme - in order to avoid key collisions in multi-theme environments - see [`VuFindTheme\Initializer#L444-L449`](https://github.com/vufind-org/vufind/blob/d4d62730859eaebbcda28ae51e1c57e4641f6e39/module/VuFindTheme/src/VuFindTheme/Initializer.php#L444-L449).

Clearly without setting a theme-specific cache only the currently enabled themes `languages` folder would be considered when flushing/rebuilding the language cache.

> it's been in the back of my mind to investigate why and whether there's a way to better break the cache when things change...

This (cache invalidation) problem can be solved in a more general way. Particularly as even more language directories might be added and because key clashes can now be avoided by using namespaces (see #1159).

We could use [the cache key](https://github.com/vufind-org/vufind/blob/cbab75ed923652d7e654a9e1067e85a8c1a9ef22/module/VuFind/src/VuFind/I18n/Initializer.php#L124) currently only computed to avoid unnecessary detection of text domains as a namespace on the translation (language) cache. This way each unique list of language directories and therefore each enabled theme containing a `languages` directory  would lead to translations being stored in a separate namespace.

Reply:
Yes, I definitely like the idea of using the cache key as a namespace to simplify some of the theme-specific caching logic. That's probably a project for a separate pull request, though, since I don't think the main idea of this PR depends on it. I can open a JIRA ticket for future reference if you don't want to deal with that right now.

Reply:
> using the cache key as a namespace

Implemented with 82116a3.

I also pushed the needed changes to theme initialization logic with 19ee9ce.